docs/user-guide.md

```diff
@@ -0,0 +1,659 @@
+If the URL is not specified we will attempt to grab the URL from the
+OpenID discovery response.
+
+## Cross-origin resource sharing (CORS)
+
+You can add a CORS header via the `--cors-[method]` with these
+configuration options.
+
+  - Access-Control-Allow-Origin
+
+  - Access-Control-Allow-Methods
+
+  - Access-Control-Allow-Headers
+
+  - Access-Control-Expose-Headers
+
+  - Access-Control-Allow-Credentials
+
+  - Access-Control-Max-Age
+
+You can add using the config file:
+
+``` yaml
+cors-origins:
+- '*'
+cors-methods:
+- GET
+- POST
+```
+
+or via the command line:
+
+``` bash
+--cors-origins [--cors-origins option]                  a set of origins to add to the CORS access control (Access-Control-Allow-Origin)
+--cors-methods [--cors-methods option]                  the method permitted in the access control (Access-Control-Allow-Methods)
+--cors-headers [--cors-headers option]                  a set of headers to add to the CORS access control (Access-Control-Allow-Headers)
+--cors-exposes-headers [--cors-exposes-headers option]  set the expose cors headers access control (Access-Control-Expose-Headers)
+```
+
+## Upstream URL
+
+You can control the upstream endpoint via the `--upstream-url` option.
+Both HTTP and HTTPS are supported with TLS verification and keep-alive
+support configured via the `--skip-upstream-tls-verify` /
+`--upstream-keepalives` option. Note, the proxy can also upstream via a
+UNIX socket, `--upstream-url unix://path/to/the/file.sock`.
+
+## Endpoints
+
+  - **/oauth/authorize** is authentication endpoint which will generate
+    the OpenID redirect to the provider
+
+  - **/oauth/callback** is provider OpenID callback endpoint
+
+  - **/oauth/expired** is a helper endpoint to check if a access token
+    has expired, 200 for ok and, 401 for no token and 401 for expired
+
+  - **/oauth/health** is the health checking endpoint for the proxy, you
+    can also grab version from headers
+
+  - **/oauth/login** provides a relay endpoint to login via
+    `grant_type=password`, for example, `POST /oauth/login` form values
+    are `username=USERNAME&password=PASSWORD` (must be enabled)
+
+  - **/oauth/logout** provides a convenient endpoint to log the user
+    out, it will always attempt to perform a back channel log out of
+    offline tokens
+
+  - **/oauth/token** is a helper endpoint which will display the current
+    access token for you
+
+  - **/oauth/metrics** is a Prometheus metrics handler
+
+## Metrics
+
+Assuming `--enable-metrics` has been set, a Prometheus endpoint can be
+found on **/oauth/metrics**; at present the only metric being exposed is
+a counter per HTTP code.
+
+## Limitations
+
+Keep in mind [browser cookie
+limits](http://browsercookielimits.squawky.net/) if you use access or
+refresh tokens in the browser cookie. Louketo Proxy divides
+the cookie automatically if your cookie is longer than 4093 bytes. The real
```

Comment:
Aside: We should deprecate splitting of cookies, it's caused us a nightmare for support with constant problems from users because of the split cookies

Reply:
@JoelSpeed I have different experience: some users may have huge tokens (for example because they have a lot of assigned groups or used IDP doesn't allow group filtering). That huge cookie with token was silently rejected by the browser (you can see set cookie in the response, but cookie was not created in the browser). That was nightmare for support - everything was configured properly, it was working for some users. But some users have had problem with infinite login loop between SP and IDP.

Reply:
There are pros and cons to split cookies. We have taken the stance that we should be logging when a cookie is too large and recommending users to use server side sessions storage when they see this log line. Eg Redis

There is also the alternative that you could return a 500 if the session is too large which would be more obvious.

There are options to explore, but I agree, silently dropping cookies is also definitely bad